adoptingerlang.org

```diff
@@ -3170,60 +3195,58 @@ In order to disable the scheduler busy waiting and to set the number of schedule
 #+END_SRC
 
-In the next section we will see how using an environment variable =${SCHEDULERS}= makes it easy to adjust the number of schedulers when we want to change the CPU limits in the Kubernetes resource.
+In the next section we will see how using an environment variable =${SCHEDULERS}= set through Kubernetes' [[https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#the-downward-api][Downward API]] makes it easy to adjust the number of schedulers when we want to change the CPU limits in the Kubernetes resource.
 
 **** Container Environments and ConfigMaps
 
-As we've seen in the Releases chapter, runtime configuration is done through environment variable substitution in =vm.args.src= and =sys.config.src=. We therefore have to insert those variable to the environment of the container. Each container in a Kubernetes Pod can have a field =env= declaring a set of environment variables:
-
-[TODO: figure out using dowardAPI for setting schedulers]
+As we've seen in the Releases chapter, runtime configuration is done through environment variable substitution in =vm.args.src= and =sys.config.src=. We therefore have to insert those variable to the environment of the container. Each container in a Kubernetes Pod can have a field =env= declaring a set of environment variables. The simplest case is explicitly providing a =name= and a =value=:
 
 #+BEGIN_SRC yaml
 env:
 - name: LOGGER_LEVEL
   value: error
-- name: PORT
-  value: 8053
-- name: SCHEDULERS
-  value: 2
 #+END_SRC
 
+This configuration would result in the environment variable =LOGGER_LEVEL= with value =error=.
+
+There are other environment variables that must be set based on the state of the container that is running. Two examples of this are setting the =NODE_IP= variable based on the IP of the Pod and setting the =SCHEDULERS= based on the CPU limits using Kubernetes' [[https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#the-downward-api][Downward API]]:
+
 #+BEGIN_SRC yaml
 env:
 - name: NODE_IP
   valueFrom:
     fieldRef:
       fieldPath: status.podIP
+- name: SCHEDULERS
+  valueFrom:
+    resourceFieldRef:
+      containerName: service-discovery
+      resource: limits.cpu
+      divisor: 1
 #+END_SRC
 
+The =fieldRef= =status.podIP= will return the current Pod's IP when it is created and the =resourceFieldRef= of =limits.cpu= with a =divisor= of =1= will return the CPU limit in a unit of CPU cores.
+
+The user defined environment variables like =LOGGER_LEVEL= can be better tracked in a Kubernetes resource specifically for configuration called a =ConfigMap=. A =ConfigMap= contains key-value pairs and can be populated from files, directories of files or literal values. Here we will use a literal value to set =LOGGER_LEVEL= to =error=:
+
 #+BEGIN_SRC yaml
 apiVersion: v1
 kind: ConfigMap
 metadata:
   name: configmap
 data:
   LOGGER_LEVEL: error
-  PORT: 8053
-  SCHEDULERS: 2
 #+END_SRC
```

Comment:
What's the difference between doing it in a configmap and just the env as earlier? Just because there's better tracking? What makes it better tracking?

Reply:
The schedulers? They are moved because it is using the "downward api" in the Deployment in order to calculate the number of schedulers based on the CPU limit instead of having to manually configure it.